docs/design/third-party-libraries.md

```diff
@@ -0,0 +1,175 @@
+doing nothing else, if necessary.
+1. Not a new package manager.
+1. Metadata storage by thirdparties in `package.json`. This must be accessible by plugins for
+the build process or even when scaffolding.
+1. Proper configuration of SystemJs in the browser.
+1. SystemJs configuration needs to be kept separated from the autogenerated part.
+1. The build process right now is faulty because the `tmp/` directory used by Broccoli is
+inside the project, which means that `node_modules/` is understood by Typescript. We need to
+proper map the imports and move them to `dist/`.
+1. Potentially hook into scaffolding.
+1. Potentially hook into the build.
+1. Safe uninstall path.
+
+# Usages
+
+Here's a few stories that should work right off:
+
+```sh
+$ ng new my-app && cd my-app/
+$ ng install jquery
+```
+^(this makes jQuery available in the index)
+
+```sh
+$ ng install angularfire2
+> Please specify your Firebase database URL [my-app.firebaseio.com]: _
+```
+^(makes firebase available and provided to the App)
+
+```sh
+$ ng install angularfire2 --dbUrl=my-firebase-db.example.com
+```
+^(skip prompts for values passed on the command line)
+
+```sh
+$ ng install angularfire2 --quiet
+```
+^(using `--quiet` to skip all prompts and use default values)
+
+```sh
+$ ng install less
+```
+^(now compiles CSS files with less for the appropriate extensions)
+
+# Proposed Solution
+
+The `install` task will perform the following subtasks:
+
+1. **Run npm install ${libName}.** On failure, fail the install process.
+1. **Run `package["angular-cli"].scripts["preinstall"]` if it exists.** If the
+script fails, run `npm uninstall ${libName}` and fail the install process.
+1. **Copy `package["angular-cli"].appData` to the `package.json` of the generated app,
+under the `"angular-cli.modules"` key.** This data will be sent to the App as is. See
+the appData section for more information.
+1. **Read the `package["angular-cli"].appPrompt` and prompt the user configuration values,
+if any.** See the appData section for more information.
+1. **Add the providers specified in `package["angular-cli"]["providers"]` to a
+`providers.js`.** See the Providers section for more information.
+1. **Run `package["angular-cli"].scripts["install"]` if it exists.** If the script fails,
+run `npm uninstall ${libName}` and fail the install process.
+1. **Detect if a package named `angular-cli-wrapper-${libName}` exist in the angular
+organization.** If so, run the steps above as if ng install angular/angular-${libName}. If
+this install fails, ignore the failure.
+
+  These packages can be used to wrap libraries that we want to support but can't update
+easily, like Jasmine or LESS.
+1. **Run `package["angular-cli"].scripts["postinstall"]` if it exists.**
+
+# Hooks
+The third party library can implement hooks into the scaffolding, and the build system. The
+CLI's tasks will look for the proper hooks prior to running and execute them.
```

Comment:
Yes. The problem is if we copy and then `ng upgrade`, we need to rebuild a lot of information.

I don't think this will be affecting performance enough to be a problem. We can revisit.